File: backend/core/src/ports/scanner.rs
//! Port scanner port (interface).

use crate::domain::PortInfo;
use crate::error::Result;

/// Port for scanning network ports.
///
/// This trait defines the interface for port scanning functionality.
/// Implementations handle platform-specific details (lsof, netstat, etc.)
pub trait PortScannerPort: Send + Sync {
    /// Scan for all listening TCP ports.
    ///
    /// Returns a list of active ports with their associated process information.
    fn scan(&self) -> impl std::future::Future<Output = Result<Vec<PortInfo>>> + Send;
}

File: cli/internal/scanner/scanner.go
package scanner

// Port represents a listening port and its associated process
type Port struct {
	Port    int    `json:"port"`
	PID     int    `json:"pid"`
	Process string `json:"process"`
	User    string `json:"user"`
	Address string `json:"address"`
	Command string `json:"command,omitempty"`
}

// Scanner interface for platform-specific implementations
type Scanner interface {
	Scan() ([]Port, error)
	Kill(pid int, force bool) error
}

// New returns a platform-specific scanner
func New() Scanner {
	return newPlatformScanner()
}
